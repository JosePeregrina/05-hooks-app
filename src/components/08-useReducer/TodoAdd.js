import React from 'react';
import useForm from '../../hooks/useForm';

export default function TodoAdd({ handleAddTodo }) {
  const [{ description }, handleFormChange, reset] = useForm({
    description: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) return;

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAddTodo(newTodo)

    reset();
  };

  return (
    <>
      <h4>Agregar Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <div className='mb-3'>
          <label htmlFor='description' className='form-label'>
            Description
          </label>
          <input
            type='text'
            name='description'
            placeholder='aprender...'
            autoComplete='off'
            className='form-control'
            onChange={handleFormChange}
            value={description}
          />
        </div>
        <div className='d-grid gap-2'>
          <button type='submit' className='btn btn-outline-primary d-md-block'>
            Agregar
          </button>
        </div>
      </form>
    </>
  );
}
