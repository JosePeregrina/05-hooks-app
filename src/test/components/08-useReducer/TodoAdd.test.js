import { shallow } from 'enzyme'
import TodoAdd from '../../../components/08-useReducer/TodoAdd'
import { demoTodos } from '../../fixures/demoTodos'

describe('The components <TodoAdd />', () => {
    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />)
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should not call the function handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit')
        const e = {preventDefault: () => {} }
        formSubmit(e)

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    })

    test('should call the function handleAddTodo with one arg', () => {
        const value = 'Aprender React'
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        })

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} })

        expect(handleAddTodo).toHaveBeenCalled()
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object))
        expect(handleAddTodo).toHaveBeenCalledWith({ 
            id: expect.any(Number),
            desc: value,
            done: false
        });
        expect(wrapper.find('input').prop('value')).toBe('')
    })
    
    
    
})
