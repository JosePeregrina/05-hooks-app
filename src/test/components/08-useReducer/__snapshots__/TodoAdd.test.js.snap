// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The components <TodoAdd /> should render correctly 1`] = `
<Fragment>
  <h4>
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="description"
      >
        Description
      </label>
      <input
        autoComplete="off"
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="aprender..."
        type="text"
        value=""
      />
    </div>
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary d-md-block"
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
